### ISSUE TITLE
Usage of @EnableJpaRepositories breaks entity scan package re-configuration

### ISSUE THREAD

**Issue by janScheible** [moduliths/moduliths#103]
Without `@EnableJpaRepositories`:
```
INFO 11928 --- [           main] figurationAndEntityScanPackageCustomizer : Re-configuring auto-configuration and entity scan packages to: com.scheible.modulith.service, com.scheible.modulith.infrastructure, com.scheible.modulith.firstfeature.
INFO 11928 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
INFO 11928 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 86ms. Found 1 JPA repository interfaces.
```

With `@EnableJpaRepositories`:
```
INFO 1872 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Bootstrapping Spring Data JPA repositories in DEFAULT mode.
INFO 1872 --- [           main] .s.d.r.c.RepositoryConfigurationDelegate : Finished Spring Data repository scanning in 98ms. Found 3 JPA repository interfaces.
INFO 1872 --- [           main] figurationAndEntityScanPackageCustomizer : Re-configuring auto-configuration and entity scan packages to: com.scheible.modulith.service, com.scheible.modulith.infrastructure, com.scheible.modulith.firstfeature.
```

With `@EnableJpaRepositories` there is an order problem. The Spring Data repositories are scanned before the re-configuration. This leads to `Caused by: java.lang.IllegalArgumentException: Not a managed type: class com.scheible.modulith.secondfeature.SecondEntity`. The issue [https://github.com/odrotbohm/moduliths/issues/4](https://github.com/odrotbohm/moduliths/issues/4) seems to be related.

The main difference I can see is that without `@EnableJpaRepositories` `RepositoryConfigurationDelegate#registerRepositoriesIn(...)` is called by `ModuleTestAutoConfiguration.AutoConfigurationAndEntityScanPackageCustomizer#registerBeanDefinitions(...)` whilst with the annoation it is called by `RepositoryConfigurationDelegate.registerRepositoriesIn(...)`.

The Spring Boot version is `2.2.5.RELEASE`. [The moduliths version was from master.](https://github.com/odrotbohm/moduliths/tree/1818a7366575c532319312db30171d7ac62205f9)

**Comment by odrotbohm** [comment]
Would you mind verifying whether that's still the case in 1.1 milestone and/or snapshots?

**Comment by janScheible** [comment]
With Spring Boot 2.4.2 and Moduliths 1.1.0-M1 it works (means no error occurs) with and without `@EnableJpaRepositories`. Thanks!

But it looks like there was a regression introduced. For me #4 strikes again now.
With Spring Boot 2.1.8, [Moduliths 1.0.0.BUILD-SNAPSHOT](https://github.com/odrotbohm/moduliths/tree/1818a7366575c532319312db30171d7ac62205f9) and without `@EnableJpaRepositories`, only the repositories for the corresponding modules were scanned, whereas with Spring Boot 2.4.2 and Moduliths 1.1.0-M1 (with and without `@EnableJpaRepositories`), always all repositories are scanned.

Started trying the previous versions and I was in luck: With Spring Boot 2.3.3 and Moduliths 1.0.0 `@EnableJpaRepositories` usage is already broken again, but the regression of #4 is also gone when omitting `@EnableJpaRepositories`.
That means  `@EnableJpaRepositories` usage was fixed in Moduliths 1.1.0-M1 but unfortunately the regression of #4 was introduced.

**Comment by odrotbohm** [comment]
Please give the latest snapshots a spin. Spring Boot 2.4 changed the way the autoconfiguration packages are handeled and our customizations were applied too late or the wrong what depending on how you look at it. That's been fixed with #153, also due with the imminent RC release.

**Comment by janScheible** [comment]
With [Moduliths 1.1.0-SNAPSHOT](64345afeac8e261486453e60715cdfd5191412de) the regression of #4 is gone. Only the relevant repositories are scanned now.
But unfortunately `@EnableJpaRepositories` annotation breaks it again with:

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'firstRepository' defined in com.scheible.modulith.firstfeature.FirstRepository defined in @EnableJpaRepositories declared on ModulithApplication: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: Not a managed type: class com.scheible.modulith.firstfeature.FirstEntity
```

**Comment by odrotbohm** [comment]
Can you summarize the required constraints and context? Does the annotation have to live in a special place to make this break?

Wouldn't mind a quick reproducer, either.

**Comment by janScheible** [comment]
I prepared a minimal example in https://github.com/janScheible/moduliths-with-enable-jpa-repositories-test.
As soon as there are 2 modules with a JPA repository the problem occurs when executing the tests.
It doesn't matter if I place the `@EnableJpaRepositories` on the application class or on a `@Configuration` class.

The test without `@ModuleTest` works. Running the application also works.

**Comment by odrotbohm** [comment]
That's very helpful, thanks. At a quick glance, the culprit is the following:

- `@EnableJpaRepositories` on the `EnableJpaRespositoryTestApplication` class *manually* enables Spring Data JPA repositories for the base package.
- `@ModuleTest` tweaks Boot's *autoconfiguration* to limit the entity scanning to the module package.

These two are obviously at odds and I haven't yet made up my mind on how the resolve that. Moving `@EnableJpaRepositories` to dedicated configuration classes in the module base packages should do the trick as a workaround.

What's the reason you manually declare `@EJR` on the root package?

**Comment by janScheible** [comment]
To be honest the sole purpose of using `@EnableJpaRepositories` was to specify `enableDefaultTransactions=false`. I was not able to find any other way (like for example a property) to set that.

I can confirm that [moving the @EnableJpaRepositories annotated @Configuration classes to the modules](https://github.com/janScheible/moduliths-with-enable-jpa-repositories-test/commit/9d31c1e73fb4d5551965171061c8f4fcdad72b69) worked.
Thank you for having a look at it.
